[WIP] Context and Multiprocessing

abdelhakeem opened the pull request:
This carries on the work done in #9943.

I've tidied the commit log ~although I do not guarantee that all commits are in a consistent state, but the current branch state is stable~ (cleaned!).

Next steps for now:

- [x] More coverage
- [x] `&:` command documentation
- [x] More examples in documentation
- [x] Context additions
- [x] Context documentation

abdelhakeem committed:
api: add nvim_grep method

abdelhakeem committed:
multiproc: add &: command

abdelhakeem committed:
multiproc: add &:[l]vimgrep[add] handlers

abdelhakeem committed:
multiproc: fix memory bugs

abdelhakeem committed:
multiproc/API: add EXEC_LUA_STATIC macro

- Added EXEC_LUA_STATIC in nvim/api/vim.h

abdelhakeem committed:
multiproc: update docs

abdelhakeem committed:
multiproc: add support for user-defined functions

abdelhakeem committed:
multiproc: update multiproc_spec.lua

- fix some memory bugs
- update multiproc_spec.lua

abdelhakeem committed:
context: return decoded values in context dict

abdelhakeem committed:
context: better error handling

abdelhakeem committed:
multiproc: more coverage

abdelhakeem committed:
multiproc: call funcs in script context (if any)

Invoke async functions in the script context in which they are
defined (if any). This allows async-called functions to use
script-local variables and functions, provided that the
corresponding context item is included in the sent context.

abdelhakeem committed:
multiproc: introduce &: default async handler

- introduce &: default async handler
- refurbish &: vimgrep async handler
- add async_handler_spec.lua

abdelhakeem committed:
multiproc: call_parallel: add "itemdone" callback

abdelhakeem committed:
multiproc: move asynccall_* functions to channel.c

The includes required for the `asynccall_*` family of
functions mess up with some unit tests on CI when included
in "nvim/channel.h", so move them to "nvim/channel.c".

abdelhakeem committed:
Context: polish context dictionary format

Make context dictionary more user-friendly:
- Eliminate buffer list unnecessary nesting
- Eliminate (type, timestamp, length) fields

abdelhakeem committed:
multiproc: fix call_parallel memory bugs

- fix call_parallel memory bugs
- improve coverage

abdelhakeem committed:
Context: fix malformed context memory leak

abdelhakeem committed:
Context: check entries in ctx_restore_funcs

abdelhakeem committed:
multiproc: improve &:vimgrep async handler

- make it non-blocking
- utilize `call_parallel` per-item callback

abdelhakeem committed:
multiproc: count AsyncCall references

abdelhakeem committed:
Context: report more info for errors

abdelhakeem committed:
Context/multiproc: save/restore script-local vars

abdelhakeem committed:
Context: handle sandboxed functions

abdelhakeem committed:
eval/multiproc: define ISLAMBDA macro

abdelhakeem committed:
Context: save/restore b:, w:, and t: vars

abdelhakeem committed:
Context: save/restore l: vars

abdelhakeem committed:
Context: save/restore parent-scope for closures

abdelhakeem committed:
multiproc: pack parent-scope l: vars for closures

abdelhakeem committed:
multiproc: set v:jobs in callbacks

- in "done" callback, v:jobs is a list of all call jobs (one element in case of `call_async`)
- in "itemdone" callback, v:jobs is a list with one element: the id of the job that completed the corresponding work item

abdelhakeem committed:
multiproc: reap async handler jobs

abdelhakeem committed:
multiproc: invoke function from within closure

Invoke functions from within a closure in the child process. This
ensures that any `l:` variables packed with the sent context are loaded
correctly and retained for later requests in case of `call_parallel`.

abdelhakeem committed:
Context: improve function handling

- Invoke functions with current_sctx.sc_sid value
- Include uf_script_ctx.sc_sid function values
- Improve error messages

abdelhakeem committed:
multiproc: integrate Context feature

abdelhakeem committed:
multiproc: improve errors + coverage

abdelhakeem committed:
Context: improve errors + coverage

abdelhakeem committed:
Context: update docs + minor changes

abdelhakeem committed:
multiproc: register closure for <SNR>_lambda_...

abdelhakeem committed:
Context: ignore "failed to write variable" errors

Silently ignore funcref vars and self-referencing containers when
writing variables

abdelhakeem committed:
Context: ignore refcounted functions

justinmk commented on src/nvim/context.h:
```diff
@@ -24,21 +25,54 @@ typedef kvec_t(Context) ContextVec;
+extern int kCtxVars;
 extern int kCtxAll;
 
+#define CONTEXT_TYPE_FROM_STR(types, str) \
```
seems like this is used in only 1 place. better to avoid macros where possible.

abdelhakeem replied:
Two places, `eval.c` and `vim.c`, in general anywhere I want to make a flags variable from some context type strings

abdelhakeem committed:
multiproc: avoid "function!" with <SNR>_lambda_{n}

Do not use a bang when redefining invoked lambda functions in the child
process so that an error is produced on name clashes with reserved names
(i.e.: "<SNR>_lambda_{n}")

abdelhakeem committed:
multiproc: add preliminary async call support

abdelhakeem commented:
I'll split this into two PR's, one for Context and one for Multiprocessing, so that the Context one can get merged soon.

justinmk commented:
Maybe useful: https://github.com/vim/vim/pull/5211 refactors `ex_vimgrep` so that the args parsing is separated.

yashpalgoyal1304 commented:
hi! what is the status of this?

abdelhakeem commented:
> hi! what is the status of this?

Hi, @yashpalgoyal1304! It's been stale for quite a long time, also not sure if there is still interest from the community...

This PR was supposed to be split into two PR's, one for the context feature and another for the multiprocessing feature. The context feature PR, which includes some improvements and some additions to the context dictionary, is at [#11029 ](https://github.com/neovim/neovim/pull/11029). The multiprocessing feature PR is yet to be opened.
